qa-tester/src/RestRoute.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * Class RestRoute.
+ *
+ * @package AmpProject\AmpWP_QA_Tester
+ */
+
+namespace AmpProject\AmpWP_QA_Tester;
+
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Server;
+
+/**
+ * Class handling the plugin's rest route.
+ *
+ * @since 1.0.0
+ */
+class RestRoute {
+
+	const REST_NAMESPACE = 'amp-qa-tester/v1';
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 */
+	public function register() {
+		add_action( 'rest_api_init', [ $this, 'register_route' ] );
+	}
+
+	/**
+	 * Registers the `/install` route that is used to install the specified plugin build.
+	 */
+	public function register_route() {
+		$route_uri = '/install';
+		$args      = [
+			'methods'             => WP_REST_Server::CREATABLE,
+			'callback'            => [ $this, 'install_callback' ],
+			'permission_callback' => static function () {
+				return current_user_can( 'update_plugins' );
+			},
+			'args'                => [
+				'id'     => [
+					'validate_callback' => static function ( $param ) {
+						if ( 'develop' === $param ) {
+							return true;
+						}
+
+						// Return true if the value matches the pattern of a release version.
+						if ( preg_match( '/^\d+\.\d+(?:\.\d+)?(?:-\w+)?$/', $param ) ) {
+							return true;
+						}
+
+						// Otherwise, validate if the value is a PR number.
+						return filter_var( $param, FILTER_VALIDATE_INT );
+					},
+					'sanitize_callback' => static function ( $param ) {
+						if ( 'develop' === $param ) {
+							return $param;
+						}
+
+						if ( preg_match( '/\d+\.\d+(?:\.\d+)?(?:-\w+)?/', $param, $matches ) ) {
+							return $matches[0];
+						}
+
+						return filter_var( $param, FILTER_SANITIZE_NUMBER_INT );
+					},
+					'required'          => true,
+				],
+
+				'origin' => [
+					'validate_callback' => static function ( $param ) {
+						return in_array( $param, [ 'release', 'branch', 'pr' ], true );
+					},
+					'required'          => true,
+				],
+
+				'url'    => [
+					'validate_callback' => 'wp_http_validate_url',
```

Comment:
Shouldn't this be a bit more strict to make sure that `url` is pointing to a ZIP file in a safe location? That it is located in our storage domain?

Reply:
Yes, good point. Will make the adjustment.

Reply:
Done in 48cd9ae368f408e305ab7ad3caad19d1bfb31c70.